core/server/worker/src/main/java/alluxio/worker/grpc/AbstractWriteHandler.java

```diff
@@ -80,65 +94,113 @@
+      mSemaphore.acquire();
+    } catch (InterruptedException e) {
+      LOG.warn("write data request {} is interrupted: {}", writeRequest, e.getMessage());
+      abort(new Error(AlluxioStatusException.fromThrowable(e), true));
+      Thread.currentThread().interrupt();
+      return;
+    }
+    mSerializingExecutor.execute(() -> {
+      try {
+        if (mContext == null) {
+          LOG.debug("Received write request {}.", writeRequest);
+          mContext = createRequestContext(writeRequest);
+        } else {
+          Preconditions.checkState(!mContext.isDoneUnsafe(),
+              "invalid request after write request is completed.");
+        }
+        if (mContext.isDoneUnsafe() || mContext.getError() != null) {
+          return;
+        }
+        validateWriteRequest(writeRequest);
+        if (writeRequest.hasCommand()) {
+          WriteRequestCommand command = writeRequest.getCommand();
+          if (command.getFlush()) {
+            flush();
+          } else {
+            handleCommand(command, mContext);
+          }
         } else {
-          handleCommand(command, mContext);
+          Preconditions.checkState(writeRequest.hasChunk(),
+              "write request is missing data chunk in non-command message");
+          ByteString data = writeRequest.getChunk().getData();
+          Preconditions.checkState(data != null && data.size() > 0,
+              "invalid data size from write request message");
+          writeData(new NioDataBuffer(data.asReadOnlyByteBuffer(), data.size()));
         }
-      } else {
-        Preconditions.checkState(writeRequest.hasChunk(),
-            "write request is missing data chunk in non-command message");
-        ByteString data = writeRequest.getChunk().getData();
-        Preconditions.checkState(data != null && data.size() > 0,
-            "invalid data size from write request message");
-        writeData(data);
+      } catch (Exception e) {
+        LogUtils.warnWithException(LOG, "Exception occurred while processing write request {}.",
+            writeRequest, e);
+        abort(new Error(AlluxioStatusException.fromThrowable(e), true));
+      } finally {
+        mSemaphore.release();
       }
-    } catch (Exception e) {
-      LogUtils.warnWithException(LOG, "Exception occurred while processing write request {}.",
-          writeRequest, e);
+    });
+  }
+
+  /**
+   * Handles write request with data message.
+   *
+   * @param request the request from the client
+   * @param buffer the data associated with the request
+   */
+  public void writeDataMessage(WriteRequest request, DataBuffer buffer) {
+    if (buffer == null) {
+      write(request);
+      return;
+    }
+    Preconditions.checkState(!request.hasCommand(),
+        "write request command should not come with data buffer");
+    Preconditions.checkState(buffer.readableBytes() > 0,
+        "invalid data size from write request message");
+    try {
+      mSemaphore.acquire();
```

Comment:
maybe there should be a helper method that does this acquire?

Reply:
Updated.